libcaf_net/test/net/web_socket/framing.cpp

```diff
@@ -0,0 +1,184 @@
+                                        pong_frame);
+        CHECK_EQ(transport->output_buffer(), pong_frame);
+      }
+    }
+    transport->output_buffer().clear();
+    WHEN("the client sends a ping with the maximum amount of data") {
+      auto data = make_test_data(125);
+      detail::rfc6455::assemble_frame(detail::rfc6455::ping, 0x0, data,
+                                      ping_frame);
+      transport->push(ping_frame);
+      CHECK_EQ(transport->handle_input(),
+               static_cast<ptrdiff_t>(ping_frame.size()));
+      THEN("the client receives a pong containing the data echoed back") {
+        detail::rfc6455::assemble_frame(detail::rfc6455::pong, 0x0, data,
+                                        pong_frame);
+        CHECK_EQ(transport->output_buffer(), pong_frame);
+      }
+    }
+  }
+}
+
+TEST_CASE("calling shutdown with protocol_error sets status in close header") {
+  CHECK_EQ(transport->start(nullptr), error{});
+  transport->configure_read(caf::net::receive_policy::up_to(2048u));
+  uut->shutdown(make_error(sec::protocol_error));
+  detail::rfc6455::header hdr;
+  detail::rfc6455::decode_header(transport->output_buffer(), hdr);
+  CHECK_EQ(hdr.opcode, detail::rfc6455::connection_close);
+  CHECK(hdr.payload_len >= 2);
+  auto error_code = (static_cast<uint16_t>(transport->output_buffer()[2]) << 8)
+                    + static_cast<uint16_t>(transport->output_buffer()[3]);
+  CHECK_EQ(error_code,
+           static_cast<uint16_t>(net::web_socket::status::protocol_error));
+}
+
+SCENARIO("the client sends an invalid ping that closes the connection") {
+  GIVEN("a valid WebSocket connection") {
+    std::vector<std::byte> ping_frame;
+    CHECK_EQ(transport->start(nullptr), error{});
+    transport->configure_read(caf::net::receive_policy::up_to(2048u));
+    WHEN("the client sends a ping with more data than allowed") {
+      auto data = make_test_data(126);
+      detail::rfc6455::assemble_frame(detail::rfc6455::ping, 0x0, data,
+                                      ping_frame);
+      transport->push(ping_frame);
+      THEN("the server rejects and closes the connection") {
+        CHECK_EQ(transport->handle_input(), 0);
+        CHECK(app->aborted);
+        CHECK_EQ(app->err, sec::protocol_error);
+        MESSAGE("Aborted with: " << app->err);
+      }
+      AND("shuts down the connection with a close header") {
+        detail::rfc6455::header hdr;
+        detail::rfc6455::decode_header(transport->output_buffer(), hdr);
+        MESSAGE("Buffer: " << transport->output_buffer());
+        CHECK_EQ(hdr.opcode, detail::rfc6455::connection_close);
+        CHECK(hdr.payload_len >= 2);
+        auto error_code
+          = (static_cast<uint16_t>(transport->output_buffer()[2]) << 8)
+            + static_cast<uint16_t>(transport->output_buffer()[3]);
+        CHECK_EQ(error_code, static_cast<uint16_t>(
+                               net::web_socket::status::protocol_error));
```

Comment:
Same here: we don't need to take care of preserving uint16.

Reply:
Right, we can use `int`. `static_cast` is still needed since we need to explicitly create `int` it from `std::byte`.

Reply:
Ah, if this is a `std::byte`: the idiomatic conversion would be [`to_integer`](https://en.cppreference.com/w/cpp/types/byte).